File: src/main/java/alfheim/common/integration/minetweaker/handler/MTHandlerSpells.java
package alfheim.common.integration.minetweaker.handler;

import alfheim.api.AlfheimAPI;
import alfheim.api.ModInfo;
import alfheim.api.spell.SpellBase;
import minetweaker.IUndoableAction;
import minetweaker.MineTweakerAPI;
import stanhebben.zenscript.annotations.ZenClass;
import stanhebben.zenscript.annotations.ZenMethod;

@ZenClass("mods." + ModInfo.MODID + ".Spells")
public class MTHandlerSpells {

	@ZenMethod
	public static void setManaCost(String name, int mana) {
		MineTweakerAPI.apply(new ManaCost(name, mana));
	}

	@ZenMethod
	public static void setCooldown(String name, int mana) {
		MineTweakerAPI.apply(new Cooldown(name, mana));
	}

	@ZenMethod
	public static void setCastTime(String name, int mana) {
		MineTweakerAPI.apply(new CastTime(name, mana));
	}

	private static class ManaCost implements IUndoableAction {

		private final SpellBase spell;
		private final int newVal;
		int oldVal = 0;
		
		public ManaCost(String name, int mana) {
			spell = AlfheimAPI.getSpellInstance(name);
			newVal = mana;
		}

		@Override
		public void apply() {
			oldVal = spell.setManaCost(newVal);
		}

		@Override
		public boolean canUndo() {
			return true;
		}

		@Override
		public void undo() {
			spell.setManaCost(oldVal);
		}

		@Override
		public String describe() {
			return String.format("Setting manacost to %d for %s", newVal, spell.name);
		}

		@Override
		public String describeUndo() {
			return String.format("Resetting manacost for %s to old value (%d)", spell.name, oldVal);
		}

		@Override
		public Object getOverrideKey() {
			return null;
		}
	}

	private static class Cooldown implements IUndoableAction {

		private final SpellBase spell;
		private final int newVal;
		int oldVal = 0;
		
		public Cooldown(String name, int mana) {
			spell = AlfheimAPI.getSpellInstance(name);
			newVal = mana;
		}

		@Override
		public void apply() {
			oldVal = spell.setCooldown(newVal);
		}

		@Override
		public boolean canUndo() {
			return true;
		}

		@Override
		public void undo() {
			spell.setCooldown(oldVal);
		}

		@Override
		public String describe() {
			return String.format("Setting cooldown to %d for %s", newVal, spell.name);
		}

		@Override
		public String describeUndo() {
			return String.format("Resetting cooldown for %s to old value (%d)", spell.name, oldVal);
		}

		@Override
		public Object getOverrideKey() {
			return null;
		}
	}

	private static class CastTime implements IUndoableAction {

		private final SpellBase spell;
		private final int newVal;
		int oldVal = 0;
		
		public CastTime(String name, int mana) {
			spell = AlfheimAPI.getSpellInstance(name);
			newVal = mana;
		}

		@Override
		public void apply() {
			oldVal = spell.setCastTime(newVal);
		}

		@Override
		public boolean canUndo() {
			return true;
		}

		@Override
		public void undo() {
			spell.setCastTime(oldVal);
		}

		@Override
		public String describe() {
			return String.format("Setting cast time to %d for %s", newVal, spell.name);
		}

		@Override
		public String describeUndo() {
			return String.format("Resetting cast time for %s to old value (%d)", spell.name, oldVal);
		}

		@Override
		public Object getOverrideKey() {
			return null;
		}
	}
}

File: src/main/java/alfheim/common/integration/minetweaker/handler/MTHandlerSpells.kt
package alfheim.common.integration.minetweaker.handler

import alfheim.api.*
import minetweaker.*
import stanhebben.zenscript.annotations.*

@ZenClass("mods." + ModInfo.MODID + ".Spells")
object MTHandlerSpells {
	
	@ZenMethod
	@JvmStatic
	fun setManaCost(name: String, mana: Int) {
		MineTweakerAPI.apply(ManaCost(name, mana))
	}
	
	@ZenMethod
	@JvmStatic
	fun setCooldown(name: String, cd: Int) {
		MineTweakerAPI.apply(Cooldown(name, cd))
	}
	
	@ZenMethod
	@JvmStatic
	fun setCastTime(name: String, time: Int) {
		MineTweakerAPI.apply(CastTime(name, time))
	}
	
	private class ManaCost(name: String, private val newVal: Int): IUndoableAction {
		
		private var spell = AlfheimAPI.getSpellInstance(name)
		internal var oldVal = 0
		
		override fun apply() {
			oldVal = spell!!.setManaCost(newVal)
		}
		
		override fun canUndo(): Boolean {
			return true
		}
		
		override fun undo() {
			spell!!.setManaCost(oldVal)
		}
		
		override fun describe(): String {
			return String.format("Setting manacost to %d for %s", newVal, spell!!.name)
		}
		
		override fun describeUndo(): String {
			return String.format("Resetting manacost for %s to old value (%d)", spell!!.name, oldVal)
		}
		
		override fun getOverrideKey(): Any? {
			return null
		}
	}
	
	private class Cooldown(name: String, private val newVal: Int): IUndoableAction {
		
		private val spell = AlfheimAPI.getSpellInstance(name)
		internal var oldVal = 0
		
		override fun apply() {
			oldVal = spell!!.setCooldown(newVal)
		}
		
		override fun canUndo(): Boolean {
			return true
		}
		
		override fun undo() {
			spell!!.setCooldown(oldVal)
		}
		
		override fun describe(): String {
			return String.format("Setting cooldown to %d for %s", newVal, spell!!.name)
		}
		
		override fun describeUndo(): String {
			return String.format("Resetting cooldown for %s to old value (%d)", spell!!.name, oldVal)
		}
		
		override fun getOverrideKey(): Any? {
			return null
		}
	}
	
	private class CastTime(name: String, private val newVal: Int): IUndoableAction {
		
		private val spell = AlfheimAPI.getSpellInstance(name)
		internal var oldVal = 0
		
		override fun apply() {
			oldVal = spell!!.setCastTime(newVal)
		}
		
		override fun canUndo(): Boolean {
			return true
		}
		
		override fun undo() {
			spell!!.setCastTime(oldVal)
		}
		
		override fun describe(): String {
			return String.format("Setting cast time to %d for %s", newVal, spell!!.name)
		}
		
		override fun describeUndo(): String {
			return String.format("Resetting cast time for %s to old value (%d)", spell!!.name, oldVal)
		}
		
		override fun getOverrideKey(): Any? {
			return null
		}
	}
}